File: api/ai/grammar.py
from http.server import BaseHTTPRequestHandler
import json
import os
import google.generativeai as genai

# Configure Google AI
api_key = os.getenv('GOOGLE_API_KEY')
if api_key:
    genai.configure(api_key=api_key)
    model = genai.GenerativeModel('gemini-2.5-flash')
else:
    model = None

class handler(BaseHTTPRequestHandler):
    def do_OPTIONS(self):
        self.send_response(200)
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'POST, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type')
        self.end_headers()
    
    def do_POST(self):
        try:
            content_length = int(self.headers['Content-Length'])
            post_data = self.rfile.read(content_length)
            data = json.loads(post_data.decode('utf-8'))
            
            text = data.get('text', '')
            if not text:
                self.send_error_response('Text is required', 400)
                return
            
            if not model:
                self.send_error_response('AI service not configured', 501)
                return
            
            prompt = f"Check the grammar and spelling of the following text and provide ONLY the corrected version without any explanations or additional text:\n\n{text}"
            corrected_text = model.generate_content(prompt)
            
            # Clean up any extra text that might come from the AI response
            import re
            corrected = corrected_text.text
            corrected = re.sub(r'^(Here\'s the corrected version:|Corrected text:|Fixed version:)\s*', '', corrected, flags=re.IGNORECASE)
            corrected = re.sub(r'^\*\*.*?\*\*:\s*', '', corrected)  # Remove **text**: patterns
            corrected = re.sub(r'^Original Text:\s*', '', corrected, flags=re.IGNORECASE)
            corrected = re.sub(r'^Corrected Text:\s*', '', corrected, flags=re.IGNORECASE)
            corrected = corrected.strip()
            
            response_data = {'corrected': corrected or text}
            self.send_success_response(response_data)
            
        except Exception as e:
            print(f"Error in grammar endpoint: {e}")
            self.send_error_response('Failed to check grammar', 500)
    
    def send_success_response(self, data):
        self.send_response(200)
        self.send_header('Content-Type', 'application/json')
        self.send_header('Access-Control-Allow-Origin', '*')
        self.end_headers()
        self.wfile.write(json.dumps(data).encode('utf-8'))
    
    def send_error_response(self, message, status_code):
        self.send_response(status_code)
        self.send_header('Content-Type', 'application/json')
        self.send_header('Access-Control-Allow-Origin', '*')
        self.end_headers()
        self.wfile.write(json.dumps({'error': message}).encode('utf-8'))

File: api/ai/grammar.js
const { GoogleGenerativeAI } = require('@google/generative-ai');

const GOOGLE_API_KEY = process.env.GOOGLE_API_KEY || 'AIzaSyCVnfhlcfFFVrRtA4tEHgCqACZX3-9QjjU';
let model = null;

try {
  if (GOOGLE_API_KEY) {
    const genAI = new GoogleGenerativeAI(GOOGLE_API_KEY);
    model = genAI.getGenerativeModel({ model: "gemini-1.5-flash" });
  }
} catch (e) {
  model = null;
}

async function callGemini(prompt) {
  try {
    if (!model) {
      const text = prompt.split('\n\n').pop() || '';
      return text;
    }
    const result = await model.generateContent(prompt);
    const response = await result.response;
    return response.text();
  } catch (error) {
    const text = prompt.split('\n\n').pop() || '';
    return text;
  }
}

module.exports = async (req, res) => {
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Methods', 'POST, OPTIONS');
  res.setHeader('Access-Control-Allow-Headers', 'Content-Type');
  
  if (req.method === 'OPTIONS') {
    res.status(200).end();
    return;
  }
  
  if (req.method !== 'POST') {
    return res.status(405).json({ error: 'Method not allowed' });
  }
  
  try {
    const { text } = req.body;
    const prompt = `Check the grammar and spelling of the following text and provide the corrected version:\n\n${text}`;
    const corrected = await callGemini(prompt);
    res.json({ corrected });
  } catch (error) {
    res.status(500).json({ error: 'Failed to check grammar' });
  }
};
